src/mapdata.h

```diff
@@ -246,16 +247,32 @@ struct map_data_common_t {
 * Short for terrain type. This struct defines all of the metadata for a given terrain id (an enum below).
 */
 struct ter_t : map_data_common_t {
-    ter_id loadid;     // This is akin to the old ter_id, however it is set at runtime.
+    ter_str_id id;    // The terrain's ID. Must be set, must be unique.
+    ter_str_id open;  // Open action: transform into terrain with matching id
+    ter_str_id close; // Close action: transform into terrain with matching id
+
     std::string trap_id_str;     // String storing the id string of the trap.
-    std::string harvestable;     // What will be harvested from this terrain?
-    std::string transforms_into; // Transform into what terrain?
-    std::string roof;            // What will be the floor above this terrain?
+    ter_str_id harvestable;     // What will be harvested from this terrain?
+    ter_str_id transforms_into; // Transform into what terrain?
+    ter_str_id roof;            // What will be the floor above this terrain?
 
     trap_id trap; // The id of the trap located at this terrain. Limit one trap per tile currently.
 
     int harvest_season; // When will this terrain get harvested?
-    int bloom_season;   // When does this terrain bloom?
+    int bloom_season;   // @todo Remove the unused field
+
+    ter_t() :
```

Comment:
This explicit constructor is probably not required. An implicit default constructor should work just fine. The ids will be default constructed, meaning they are an empty string, which is fine. They are overridden by the loading code anyway. And if they aren't properly loaded, the final consistency checks should recognize the empty string ids as invalid and complain about that.